webapp/client/src/pages/TaxNumberPage.js

```diff
@@ -0,0 +1,331 @@
+  TaxOfficeSelection: "TaxOfficeSelection",
+  TaxNumberInput: "TaxNumberInput",
+  RequestNewTaxNumber: "RequestNewTaxNumber",
+};
+
+function currentState(taxNumberFormState, fields) {
+  if (taxNumberFormState === TAX_NUMBER_FORM_STATES.NothingSelected) {
+    if (fields.steuernummerExists.value === "yes") {
+      if (fields.bundesland.selectedValue) {
+        return TAX_NUMBER_FORM_STATES.TaxNumberInput;
+      }
+      return TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists;
+    }
+    if (fields.steuernummerExists.value === "no") {
+      if (fields.bundesland.selectedValue) {
+        if (fields.bufaNr.selectedValue) {
+          return TAX_NUMBER_FORM_STATES.RequestNewTaxNumber;
+        }
+        return TAX_NUMBER_FORM_STATES.TaxOfficeSelection;
+      }
+      return TAX_NUMBER_FORM_STATES.StateSelectionNoTaxNumberExists;
+    }
+  }
+
+  return taxNumberFormState;
+}
+
+function extractCorrespondingTaxOffices(
+  selectedStateAbbreviation,
+  taxOfficeList
+) {
+  const selectedState = taxOfficeList.find(
+    (taxOffice) => taxOffice.stateAbbreviation === selectedStateAbbreviation
+  );
+
+  if (selectedState) {
+    const selectedTaxOffices = selectedState.taxOffices;
+    if (selectedTaxOffices) {
+      selectedTaxOffices.sort((a, b) => String(a.name).localeCompare(b.name));
+      return selectedTaxOffices.map((taxOffice) => ({
+        value: taxOffice.bufaNr,
+        displayName: taxOffice.name.replace("Finanzamt ", ""),
+      }));
+    }
+  }
+
+  return [];
+}
+
+export default function TaxNumberPage({
+  stepHeader,
+  form,
+  fields,
+  taxOfficeList,
+  numberOfUsers,
+}) {
+  const { t } = useTranslation();
+
+  const [taxNumberFormState, setTaxNumberFormState] = useState(
+    TAX_NUMBER_FORM_STATES.NothingSelected
+  );
+
+  const [selectedStateAbbreviation, setSelectedStateAbbreviation] = useState(
+    () => {
+      if (fields.bundesland.selectedValue) {
+        return fields.bundesland.selectedValue.toLowerCase();
+      }
+      return undefined;
+    }
+  );
+
+  const changeTaxNumberExists = (event) => {
+    switch (taxNumberFormState) {
+      case TAX_NUMBER_FORM_STATES.NothingSelected:
+      case TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists:
+      case TAX_NUMBER_FORM_STATES.StateSelectionNoTaxNumberExists:
+        if (event.target.value === "yes") {
+          setTaxNumberFormState(
+            TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists
+          );
+        } else if (event.target.value === "no") {
+          setTaxNumberFormState(
+            TAX_NUMBER_FORM_STATES.StateSelectionNoTaxNumberExists
+          );
+        }
+        break;
+      default:
+        if (event.target.value === "yes") {
+          setTaxNumberFormState(TAX_NUMBER_FORM_STATES.TaxNumberInput);
+        } else if (event.target.value === "no") {
+          setTaxNumberFormState(TAX_NUMBER_FORM_STATES.TaxOfficeSelection);
+        }
+        break;
+    }
```

Comment:
>Does this mean that in all three states (ll. 91-93) the same code is executed (ll.94-103)?

Yes. What helped me to get breaks was to think of the different cases as rooms on top of each other. If there is no break, there is no floor. If you go into a room, you will fall down until you reach a floor (break).

> What do we then need the default for? I think I'd prefer to list all states explicitly here. Same with the other switch cases.
I did not want to differentiate between all the other cases (is the tax offices already selected or not) because we just want to set it to one of these two states (taxNumerInput or TaxOfficeSelection) then.

Reply:
Yes it makes more sense like this and I actually changed my opinion on this. 👌